### DISCUSSION TITLE
Add an alias for the primary constructor

### DISCUSSION THREAD

**Discussion by Mr0N** [dotnet/csharplang#7720] [Language Ideas]
I liked the primary constructor more or less, but there is a problem that the external API depends on the internal implementation. That is, if you change the name of the constructor parameters, the entire logic of the program will break. I suggest adding aliases for the constructor parameters.

```
 public class MoveGeoResult(Config config alias _config, Node node alias _node) : IMoveGeoResult
 {
     public void MoveGeo()
     {
      
     }

 }
```

**Comment by HaloFour** [comment]
> if you change the name of the constructor parameters, the entire logic of the program will break.

How does that make it different from any other identifier in C#?

**Reply by Mr0N** [reply]
Because this is a public api

**Reply by Mr0N** [reply]
![image](https://github.com/dotnet/csharplang/assets/61237140/197fd55c-8189-49cc-86fd-5b0c0f5b5e79)

**Reply by HaloFour** [reply]
So are parameters to literally any method, and changing their names could always be a source breaking change.

**Reply by Mr0N** [reply]
In methods, these names are used only inside the method, these parameters are used in the entire class. Changing the name of the parameter affects only the method, and here the entire class, if the class is large, then changing the name of the parameter can break the entire class.

**Reply by Mr0N** [reply]
Previously, either underlining or this was used for "this"

**Reply by HaloFour** [reply]
> In methods, these names are used only inside the method, these parameters are used in the entire class. Changing the name of the parameter affects only the method, and here the entire class, if the class is large, then changing the name of the parameter can break the entire class.

So then don't do that.

I understand your intent is to try to name them differently from the parameter so that you can treat them as fields.  If you want to do that, assign them to fields and don't reference the capture.  The compiler will warn you if you attempt to reference the capture again beyond the lifetime of initialization.

**Reply by Mr0N** [reply]
Currently "this" is not available, and _ would look bad for a public api

**Reply by Mr0N** [reply]
Not so many people look at compiler warnings.

**Reply by HaloFour** [reply]
> Not so many people look at compiler warnings.

That sounds like their problem.  The warning can be promoted to an error if someone wants to ensure that it breaks compilation.

**Reply by Mr0N** [reply]
well, I don't know, it's easier to add aliases, there will be fewer problems, and the constructor will not violate encapsulation

**Reply by HaloFour** [reply]
> it's easier to add aliases

Adding language features is not easy.

> and the constructor will not violate encapsulation

There is no violation of encapsulation.  To the outside world the parameter names are the only thing that matters.  Consumers aren't even aware that primary constructors exist.  How the parameters are used within the type are entirely implementation details.

**Reply by Mr0N** [reply]
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Collections/src/System/Collections/BitArray.cs
For large classes, the primary constructor turns into a disaster

**Reply by HaloFour** [reply]
Primary constructors aren't intended to be used with all classes, nor are they intended to replace normal constructor syntax.  They are an opinionated short-hand for simple scenarios.  `BitArray` is definitely not one of those scenarios.

**Comment by Mr0N** [comment]
That is, the main name is displayed for the main user, and aliases are available only for the implementation of the class
